### ISSUE TITLE
Volume numbers with letter suffix break the left pane volume list behavior when viewing an article

### ISSUE THREAD

**Issue by nrshapiro** [Psychoanalytic-Electronic-Publishing/PEP-Web-User-Interface#540]
In PEP IDs, PEP uses the suffix "S" after a volume number (example: APA 27S, IJP 34S) to indicate a supplemental volume, or A-Z to indicate the issue number (but only when issues repeat numbering).  This is always how we've done it, and it unfortunately means you need to do some exception handling to deal with the variability. 

![image](https://user-images.githubusercontent.com/4837554/125100482-259ae300-e0a7-11eb-919c-be88ddc46a10.png)

And here, unfortunately, this has caused some problem behavior in the client. 

For example, browse Volume 34 of IJP.  Then click on an article from one of the regular issues, you see the article, and the list of articles from Vol 34 on the left side pane.  The client does a search sending the following parameter in facetquery:

'art_sourcecode:(IJP) AND art_vol:(34)'

However, go back to the contents list for Vol 34, and scroll down to the supplemental issue, and click one of those articles.  There I think it's picking up the volume number from the XML, and it sees the 34S, and sends this query instead:

art_sourcecode:(IJP) AND art_vol:(NaN)

which produces, no articles, and so the left hand pane lists no articles.

This is also a problem for journals with repeating page numbers across issues--the ones we use a letter suffix A-Z for the volume number.

Browse to AIM, select Volume 1, and then view an article in issue 2, for example.  Unlike other "regular" articles without suffixes in their volume number, no articles will be shown on the left.

While I admit the use of the suffix on the volume number is the cause of the problem, there are reasons we code like that. The prime one being that our original journals used continuous numbering across a volume, so including issue was not necessary in the IDs.  But supplements can start with page 1, and also now we have some journals that restart page numbering each issue.  And others that have done that inconsistently over the years.

I will provide further information documenting the interactions and codes in comments below this initial one.

**Comment by nrshapiro** [comment]
@bakerac4 

I am making some changes to the API to make it easier to understand IDs, and to use them.  Of course, sending an ID to Smarttext works in all cases but that's for a single article.  

But if you are querying using Solr syntax, as you appear to be doing in this case, using parameter facetquery to accomplish the left pane article search, you're on your own and need to understand the schema and build queries accordingly.   In the stored Solr data, Supplements are currently indicated in the solr art_iss column as "Supplement", and art_vol never includes the volume suffix, e.g., 34S.

**Solution:** You should be using the Search parameters to do the search.  This would not have completely fixed this before, but in the latest version of the API (2021.0714/v2.1.37, on Stage), although suffixes are not part of the data stored in the issue and volume fields, the server will now handle them.  For example:

- If you put 33C in the volume parameter, the server will interpret this as Volume 33 Issue 3.  
- If you put 34S in the volume parameter, it will do a search for Volume 33 Issue Supplement

Since this is all being done in the server and it's configurable to the data, it will be easy to change later.

Also, by way of documenting and helping API users to better understand the articleIDs, I've added a new endpoint, which will take a PEP article ID and break it down into the correct parameter data.  

e.g., IJP.034S.0118A 

Returns:
{
  "articleID": "IJP.034S.0043A",
  "standardized": "IJP.034S.0043A",
  "isArticleID": true,
  "sourceCode": "IJP",
  "volumeSuffix": "S",
  "volumeInt": 34,
  "volumeNbrStr": "034",
  "issueCode": "S",
  "isSupplement": true,
  "issueInt": 0,
  "pageNbrStr": "0043",
  "pageInt": 43,
  "isRoman": false,
  "pageSuffix": "A",
  "articleInfo": null,
  "allInfo": false
}

and IJP.101A.R0001A

{
  "articleID": "IJP.101A.R0001A",
  "standardized": "IJP.101A.R0001A",
  "isArticleID": true,
  "sourceCode": "IJP",
  "volumeSuffix": "A",
  "volumeInt": 101,
  "volumeNbrStr": "101",
  "issueCode": "A",
  "isSupplement": false,
  "issueInt": 0,
  "issueCodeInt": 1,
  "pageNbrStr": "0001",
  "pageInt": 1,
  "romanPrefix": "R",
  "isRoman": true,
  "pageSuffix": "A",
  "articleInfo": null,
  "allInfo": false
}

This does not currently check whether that information really maps to an article, but it still should correctly take apart any ID.  I'm still refining it so I may add a field like "exists".

Let me know if you have questions.  I'm not sure why you are using facetquery here, but if you use the Search parameters instead for volume and/or issue, I think that will fix the problem.

When we stop publishing to GVPi's server, it will be easier for me to see what I can do about cleaning up inconsistencies of our ID and naming system (it's a result of years of adapting to the highly variable way the journals publish--even with a journal).

**Comment by bakerac4** [comment]
@nrshapiro it would be quick for me to just handle having a string there. Which I have done, but Im still seeing 0 results returned.

```https://stage-api.pep-web.rocks/v2/Database/Search/?abstract=false&facetquery=art_sourcecode%3A(IJP)+AND+art_vol%3A(034S)&formatrequested=XML&limit=40&offset=0&sort=bibliographic&synonyms=false```

**Comment by nrshapiro** [comment]
@bakerac4 That's because 34S isn't an issue, it's a volume that has an issue suffix.  So you put it in the volume parameter.

The changes are: 
 - You can add a volume suffix (issue code) to the volume parameter, and it parses it into volume and issue number
 - You can use an alpha code (A-Z) in the issue parameter, and it translates it to either an issue number or the name Supplement.

i.e., 34S in the volume parameter get's parsed out to Vol 34, issue Supplement

The only change to the issue parameter is that it accepts issue codes.  It does not accept volume numbers, or volume numbers + issue codes.

So in other words:
- 34 in the volume parameter and S in the issue parameter, would work.  It can even work if you mistakenly put 34S in the volume parameter and S in the issue parameter.
- 34C in the volume parameter would give you Vol 34 and Issue 3, since A=issue 1, B=issue 2, etc.

These parameters have never expected parentheses around them, so that will throw them off, though I can deal with that in a later version.

**Comment by nrshapiro** [comment]
@bakerac4  DOUBLE THAT SORRY.  LET ME START OVER.  

I analyzed your string wrong.

Your problem is that you are still sending it to the facetquery parameter.  You are not using the built in search.

https://stage-api.pep-web.rocks/v2/Database/Search/?abstract=false&art_sourcecode%3A(IJP)+AND+art_vol%3A(034S)&formatrequested=XML&limit=40&offset=0&sort=bibliographic&synonyms=false

**Comment by nrshapiro** [comment]
@bakerac4 Putting this in a new message to make sure you get notified.

Using the interactive documentation, it's easy to test these things.  Better than editing a URL (at least for me).

https://stage-api.pep-web.rocks/v2/Database/Search/?synonyms=false&volume=34S&viewperiod=4&abstract=false&formatrequested=HTML&similarcount=0&highlightlimit=5&facetmincount=1&facetlimit=15&facetoffset=0&sort=score%20desc&limit=15&offset=0

That works correctly, but there's another problem...the lead 0 is throwing off the server if you do it that way.  I'm not sure why yet, but should be easy to fix.

https://stage-api.pep-web.rocks/v2/Database/Search/?synonyms=false&volume=034S&viewperiod=4&abstract=false&formatrequested=HTML&similarcount=0&highlightlimit=5&facetmincount=1&facetlimit=15&facetoffset=0&sort=score%20desc&limit=15&offset=0

Currently, this returns all records.

Parens will also throw it off and it will return all records.  There's no need for parens there, but I can see about removing them in the server.

Note in my example, I didn't bother with the IJP param.

**Comment by nrshapiro** [comment]
Yes.

Here's the skinny on that parameter:

1) The volume + issue code (e.g., 34S) parsing is currently set in the opas configuration file to make it easy to adapt to other coding systems.  At least that's how I was thinking of it when adding it a few days ago, but to be that generalized, I'm sure there will be more to do elsewhere.
2) The Volume (and some other) parameters allow Booleans.   E.g., you could put [33 OR 34](https://stage-api.pep-web.rocks/v2/Database/Search/?synonyms=false&volume=34%20OR%2033&viewperiod=4&abstract=false&formatrequested=HTML&similarcount=0&highlightlimit=5&facetmincount=1&facetlimit=15&facetoffset=0&sort=score%20desc&limit=15&offset=0) in that field and get both volumes.  But I adopted the volume/issue parser as a quick fix to allow you to ask for a volume and issue code, without worrying about what it means. So currently, it just looks to see if there's a Boolean in the field, and if there is, it skips the volume issue parse.
3) Parens can be used in the Boolean parse case, those were already handled.  I've added them in the single use case as well, though it's just going to remove them.
4) Leading 0's fail because a) vol is a numeric string field, and 034 is not what's stored in Solr, so the search fails to find vol 34.  I've now stripped leading zeros as well as parens.  But again, not in the case where Booleans are used because I don't have a full parse implemented that looks at all the subparts, and it's also further downstream of the new code to handle volume/issue parsing.  Maybe someday.  But I'm more inclined to later try and make the use of volumes and issue codes completely consistent (no defaults).

I still have to finish running tests and then I'll load to the server.  I will also test to make sure other parameters, where possible, ignore outer parens, if they don't already.  But that won't be this next version.

**Comment by bakerac4** [comment]
Ok, I have fixed this in #546 . Feel free to test it out and let me know, or just let me know and Ill get it reviewed and merged.

**Comment by nrshapiro** [comment]
@bakerac4 I just loaded the client from your fix locally, but let it run the server from stage, since it's been updated here.

- I was able to see the 5 papers in IJP.034S in the left hand sidebar when I browsed to the TOC, and selected one of the papers.
      http://stage.pep-web.rocks/browse/document/IJP.034S.0043A?page=P0043
- I was able to see the papers in the left hand sidebar of Vol 3 issue 4 of FA when I selected a paper from that issue.  (FA is one of the vols that uses letters for issue numbers in the ID because they restart page numbers every issue.)
       https://stage.pep-web.rocks/browse/document/FA.003C.0323A?page=P0323

The above links won't work online until you merge and we load the new client, since my tests used localhost.  Go ahead and merge it!  Thanks.
